pkg/metrics/events.go

```diff
@@ -0,0 +1,159 @@
+package metrics
+
+// This file contains wrapper functions that will map a Prometheus metric names to its
+// label field, help field and an associated Event.
+
+// MetricType will represent a Prometheus metric.
+// A variable of type MetricType will hold the name of the Prometheus metric as reported.
+type MetricType string
+
+const (
+	SampleCountType MetricType = "sample_count"
+
+	ActionSetBackupCreatedType   MetricType = "actionset_backup_created_count"
+	ActionSetBackupCompletedType MetricType = "actionset_backup_completed_count"
+
+	ActionSetRestoreCreatedType   MetricType = "actionset_restore_created_count"
+	ActionSetRestoreCompletedType MetricType = "actionset_restore_completed_count"
+
+	ActionSetTotalCreatedType   MetricType = "actionset_total_created_count"
+	ActionSetTotalCompletedType MetricType = "actionset_total_completed_count"
+)
+
+// MetricTypeOpt is a struct for a Prometheus metric.
+// Help and LabelNames are passed directly to the Prometheus predefined functions.
+// EventFunc holds the constructor of the linked Event of a given MetricType.
+type MetricTypeOpt struct {
+	EventFunc  interface{}
+	Help       string
+	LabelNames []string
+}
+
+// Mapping a Prometheus MetricType to the metric MetricTypeOpt struct.
+// Basically, a metric name is mapped to its associated Help and LabelName fields.
+// The linked event function (EventFunc) is also mapped to this metric name as a part of MetricTypeOpt.
+var MetricTypeOpts = map[MetricType]MetricTypeOpt{
+	SampleCountType: {
+		EventFunc:  NewSampleCount,
+		Help:       "Sample counter to remove later",
+		LabelNames: []string{"sample"},
+	},
+
+	ActionSetBackupCreatedType: {
+		EventFunc: NewActionSetBackupCreated,
+		Help:      "The count of backup ActionSets created",
+	},
+	ActionSetBackupCompletedType: {
+		EventFunc:  NewActionSetBackupCompleted,
+		Help:       "The count of backup ActionSets completed",
+		LabelNames: []string{"state"},
+	},
+
+	ActionSetRestoreCreatedType: {
+		EventFunc: NewActionSetRestoreCreated,
+		Help:      "The count of restore ActionSets created",
+	},
+	ActionSetRestoreCompletedType: {
+		EventFunc:  NewActionSetRestoreCompleted,
+		Help:       "The count of restore ActionSets completed",
+		LabelNames: []string{"state"},
+	},
+
+	ActionSetTotalCreatedType: {
+		EventFunc:  NewActionSetTotalCreated,
+		Help:       "The count of total ActionSets created",
+		LabelNames: []string{"actionType", "namespace"},
+	},
+	ActionSetTotalCompletedType: {
+		EventFunc:  NewActionSetTotalCompleted,
+		Help:       "The count of total ActionSets completed",
+		LabelNames: []string{"actionType", "namespace", "state"},
+	},
+}
+
+// Event describes an individual event.
```

Comment:
It looks like much of this was modeled after the counter metrics I added in K10, so while I can't speak much to this code, I can offer some insights into the K10 design.

There were two primary things considered in K10's design:

1. How can we leverage the type system to prevent the Prometheus library panicking unexpectedly?
2. How can we model the metrics to conceptually match what's going on in the system?

---

The Prometheus library has an annoying propensity to panic if a `CounterVec` is requested with a different set of labels than it was instantiated with. I ran into this several times while prototyping the solution, so I decided to employ the type system to help protect against this.

This meant that the developer shouldn't be able to freely associate a metric name with labels (because that was the case where we kept running into trouble and causing panics).

So the metric type had a private metric name and set of labels that were instantiated via a constructor. This meant that we only needed to ensure that the constructors worked properly, which we did through a reflection based unit test.

---

As for conceptual models for the metrics, I elected to use "event" over "sample" because inherently the transitionary event between a previous state and a new state (in other words, it's representing the point at which an increase happens).

This is distinct from what I think of as a "sample": the measure of the overall state at an instantaneous point in time (_not_ the transition between two states, just the measure of the state).

To put a different way, I thought of a "sample" more as reporting the current value (i.e. a gauge), whereas "events" were the accumulation of individual changes (i.e. a counter).

---

So if we're going to eschew the type protections, we should recognize what the consequences might be and we should decide if the terminology aligns with conceptual models (personally, "sample" doesn't seem to make as much sense to me in this context).

Reply:
Before we toss away the type system prevention of Prometheus panics, please note that the Event structure facilitates writing a unit test that validates that the labels from each constructor match the labels that are used to initialize the CounterVec. Or so I've been told :).

Reply:
I don't think I am advocating that we tossed away the type system. In my diff, there is still a constructor to constrain the labels that can be passed in. E.g.,

```go
func NewSampleActionSetTotal(action string, namespace string) *Sample {
   // ....
}
```

It isn't obvious to me how my diff is hindering the type system protection you are referring to. My intention is still this:

> What do you think of merging Event, MetricType and MetricTypeOpt into a Metric and Sample structs?

Re: conceptual model, I picked "sample" because it represents a data point instance in the timeseries. Yeah, I agree it doesn't align as well as I'd want, because in a TSDB, a sample of a metric time series is a k/v pair of timestamp to value (with all the label values defined). In this case, we only have the label values, not its value at that given time. Hence, I said, "_closer to what you are trying to describe_" 🙂. What I have in mind is like `MetricData`, `MetricDataPoint`(not saying those should be the name).

I am not convinced "event" is clearer; feels like this is where it gets subjective. I think we agree that type system protection is good, and we want some way to conceptually represent the metric and distinguish it from the metric instance at a given time, right?

Reply:
Perhaps the concern with your diff is that the `Name` and `Labels` are public fields. Wouldn't we be worried about developers constructing `Sample`s inline and not using the constructors, thus bypassing the type system?

--

I agree that I'm not convinced "event" is more clear and that it's a very subjective topic 😄 However, one additional consideration might be: is using "sample" here going to cut off the term we would want to use for gauges in the future (when we add gauge metrics)?

Reply:
Unexporting the public fields sound actionable to me. Personally, I am not that worried about the panics caused by initialization with unknown labels :smile:, because:  

1. I am not convinced that they won't be caught during developer testing and code review (assuming they happen :smile:)
2. Most common vectors methods like `With()` and `WithLabelValues()` that panic, have a non-panic `Get*` counterpart

I expect developers to learn and understand the packages they are using (through docs, code review, mentorship etc.), instead of relying on others to fail-proof things for them.

I am not opposed to unexporting the public fields though.

As for the usage of "sample", yeah, it's not a good choice. I prefer not to overload the TSDB definition of "sample", which is a k/v pair of timestamp to values, regardless of the metric type. I don't have a better name for it, so `Event` is fine (it's more generic than "sample") until we can come up with something else.